File: app/day06/parts.ml
open Support.Utils

let file = "day06/input.txt"


let rec zip lst1 lst2 = 
  match lst1, lst2 with
  | x::xs, y::ys -> (x, y) :: zip xs ys
  | _ -> []

let rec range start finish =
  if start > finish then []
  else start :: range (start + 1) finish
;;

let parse_line_to_num (line: string) (clear_str: string) =
  Str.(global_replace (Str.regexp clear_str) "" line) 
  |> String.split_on_char ' ' 
  |> List.filter ((<>) "") 
  |> List.map int_of_string
;;

let parser_part_1 (lines: string list) = 
  match lines with
  | [time; distance] ->
    let t = parse_line_to_num time "Time: " in
    let d = parse_line_to_num distance "Distance: " in
    zip t d
  | _ -> failwith "Invalid input"
;;

let _compute (lines: string list) = 
  parser_part_1 lines
  |> List.map (fun (t, d) -> 
    range 0 t
    |> List.fold_left (
      fun acc c -> if ((t - c) * c > d) then acc + 1 else acc
    ) 0
  )
;;

let () = 
  let lines = read_lines file in
  let result = _compute lines
  |> List.fold_left (fun acc cur -> acc * cur) 1 in
  string_of_int result |> print_endline;
  assert (result == 288)
;;

File: app/day06/parts.js
const { readFile, strToInt } = require("../support/utils");

const parse_line_to_num = (line, clear_str) =>
  line
    .replace(clear_str, "")
    .split(" ")
    .filter((n) => n)
    .map((n) => strToInt(n));

const parser_part_1 = (lines) => {
  const [time, distance] = lines;
  const t = parse_line_to_num(time, "Time: ");
  const d = parse_line_to_num(distance, "Distance: ");
  return t.map((t, i) => [t, d[i]]);
};

const main = () => {
  const lines = readFile("day06/input.txt").split("\n");

  const races = parser_part_1(lines);
  const results = races.map(([t, d]) => {
    return Array.from(Array(t).keys()).reduce((acc, c) => {
      if ((t - c) * c > d) return acc + 1;
      return acc;
    }, 0);
  });
  return results.reduce((a, b) => a * b, 1);
};

console.info(main());
